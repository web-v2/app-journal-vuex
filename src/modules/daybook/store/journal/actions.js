import journalApi from "@/api/journalAPI"
/*
export const myAction = async ({ commit }) => {

}
https://vue-demos-v2-default-rtdb.firebaseio.com/entries.json
*/

export const loadEntries = async ({ commit }) => {
    const {data} = await journalApi.get('/entries.json')
    if(!data){
        commit('setEntries', [])
        return
    }
    const entries = []
    for (let id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id]
        })
    }
    commit('setEntries', entries)
}

export const updateEntry = async ({ commit }, entrada) => {
    const {fecha, picture, text} = entrada
    const dataToSave = {fecha, picture, text}
    const resp = await journalApi.put(`/entries/${entrada.id}.json`, dataToSave)    
    commit('updateEntry', {...entrada}) //commit updateEntry en Mutaciones para actualizar el state.
}

export const createEntry = async ({ commit }, entrada) => {
    const {fecha, picture, text} = entrada
    const dataToSave = {fecha, picture, text}
    //console.log(dataToSave);
    const {data} = await journalApi.post(`/entries.json`, dataToSave)    
    dataToSave.id = data.name;
    commit('addEntry', dataToSave)
    return data.name;
}

export const deleteEntry = async ({ commit }, id) => {
    const resp = await journalApi.delete(`/entries/${id}.json`)    
    commit('deleteEntry', id) //commit deleteEntry en Mutaciones para actualizar el state.
    return resp
}