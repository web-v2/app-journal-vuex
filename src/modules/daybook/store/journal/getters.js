/*
export const myGetter = ( state ) => {

    return state

}
*/

export const getEntriesByTerm = ( state ) => ( term = '') =>{
    if ( term.length === 0 ) return state.entries

    return state.entries.filter( entry => entry.text.toLowerCase().includes( term.toLocaleLowerCase() ))
}

export const getEntryById = ( state ) => (id = '') => {
    const ent = state.entries.find( entry => entry.id === id );
    if(!ent) return
    
    return { ...ent }
}