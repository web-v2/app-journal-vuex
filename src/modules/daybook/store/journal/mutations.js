/*
export const myMutation = ( state ) => {

}
*/
export const setEntries = ( state, entries ) => {
    state.entries = [...state.entries, ...entries];
    state.isLoading = false;
}

export const updateEntry = ( state, entry ) => {
    const idx = state.entries.map(e => e.id).indexOf(entry.id);
    state.entries[idx] = entry;
}

export const addEntry = (state, entry ) => {
    state.entries = [entry, ...state.entries];
}

export const deleteEntry = (state, id ) => {
    state.entries = state.entries.filter(e => e.id != id);
}