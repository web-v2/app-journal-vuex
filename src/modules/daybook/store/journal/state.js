//https://vue-demos-v2-default-rtdb.firebaseio.com/entries.json
export default () =>({
    isLoading: true,
    entries: []
})