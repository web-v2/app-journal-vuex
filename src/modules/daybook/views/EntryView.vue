<template>
    <template v-if="entry">
        <div>
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{day}}</span>
                <span class="mx-1 fs-3">{{month}}</span>
                <span class="mx-2 fs-4 fw-light">{{yearDay}}</span>
            </div>
            <div>
                <input type="file" name="carga" @change="onSelectedImagen" ref="imageSelector" v-show="false" accept="image/png, image/jpeg">
                <button class="btn btn-danger mx-2" v-if="entry.id" @click="onDeleteEntry">Borrar <i class="fa fa-trash-alt"></i></button>
                <button class="btn btn-primary" @click="onSelectImagen">Subir foto <i class="fa fa-upload"></i></button>
            </div>
        </div> 
        <hr>
        <div v-if="entry" class="d-flex flex-column px-3 h-75">
            <textarea name="post" id="post" cols="10" rows="3" placeholder="¿Qué sucedió hoy?" v-model="entry.text"></textarea>
        </div>            
        </div>
        <img 
            v-if="entry.picture && !localImage"
            :src="entry.picture"
            alt="entry-pinture"
            class="img thumbnail"
        >
        <img 
            v-if="localImage"
            :src="localImage"
            alt="entry-pinture"
            class="img thumbnail"
        >  
    </template>
    <Fab icon="fa-save" @on:click="saveEntry"/>
</template>
<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters, mapActions } from 'vuex'
import Swal from 'sweetalert2'
import getDayMonthYear from '../helpers/getDayMonthYear'
import uploadImage from '../helpers/uploadImage'

export default {
    name: 'EntryView',
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(()=>import('../components/Fab.vue'))
    },
    data(){
        return {
            entry: null,
            localImage: null,
            fileGeneral: null
        }
    },
    computed:{
        ...mapGetters('journal', ['getEntryById']),
        day(){
            const { day } = getDayMonthYear( this.entry.fecha )
            return day
        },
        month(){
            const { month } = getDayMonthYear( this.entry.fecha )
            return month
        },
        yearDay(){
            const { yearDay } = getDayMonthYear( this.entry.fecha )
            return yearDay
        },                
    },
    methods:{
        ...mapActions('journal',['updateEntry','createEntry','deleteEntry']),

        loadEntry(){
            let ent;
            if( this.id === 'new' ){
                ent = {
                    text : '',
                    fecha : new Date().getTime()
                }
            }else{
                ent = this.getEntryById( this.id )
                if(!ent) return this.$router.push({ name: 'no-entry' })
            }
            this.entry = ent
        },

        async saveEntry(){   
            new Swal({
                title : 'Espere por favor...',
                AllowOutsideClick : false
            })         
            Swal.showLoading()
            
            //Carga de imagen
            const picture = await uploadImage(this.fileGeneral)            
            this.entry.picture = picture

            if(this.entry.id){
                //Actualizar
                await this.updateEntry(this.entry); //updateEntry en Store/app-journal/Actions
            }else{
                //Crear nueva entradas
                const idNewReg = await this.createEntry(this.entry);
                this.$router.push({ name: 'entry', params: {id: idNewReg}})
            }
            this.fileGeneral = null
            Swal.fire('Guardado', 'Entrada registrada con éxito!', 'success')
        },
        async onDeleteEntry(){              
            const {isConfirmed} = await Swal.fire({
                title: '¿Está seguro?',
                text: 'Una vez borrado, no se puede recuperar',
                showDenyButton: true,
                confirmButtonText: 'Si, estoy seguro!'                
            })              
            if(isConfirmed){
                Swal.fire({
                    title : 'Espere por favor...',
                    AllowOutsideClick : false
                }) 
                Swal.showLoading()                
                await this.deleteEntry(this.entry.id);                
                this.$router.push({name: 'no-entry'})
                Swal.fire('Eliminado', '', 'success')
            }     
        },
        onSelectedImagen(event){
            const file = event.target.files[0];
            if(!file){
                this.localImage = null
                this.fileGeneral = null
                return
            }
            this.fileGeneral = file
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL( file )
        },
        onSelectImagen(){
          this.$refs.imageSelector.click()
        }
    },
    created() {
        this.loadEntry()
    },
    watch:{
        id(){
            this.loadEntry()
        }
    }
}
</script>

<style lang="scss" scoped>
    textarea{
        font: size 20px;
        border: none;
        height: 100%;

        &:focus{
            outline: none;
        }
    }
    img{ 
        width: 200px;
        position: fixed;
        bottom: 150px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);        
    }
</style>
