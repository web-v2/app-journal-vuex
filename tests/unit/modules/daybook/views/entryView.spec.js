import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import Swal from 'sweetalert2'
import journal from '@/modules/daybook/store/journal'

import { journalState } from '../../../mock-data/test-journal-state'
import EntryView from '@/modules/daybook/views/entryView.vue'
const createVuexStore = ( initialState ) => 
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initialState }
            }
        }
    })

    jest.mock('sweetalert2', () => ({
        fire: jest.fn(),
        showLoading: jest.fn(),
        close: jest.fn()
    }))

describe('Pruebas en el EntryView', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }
    
    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: '-NH7uKXKp-qWoAlduDNp'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })

    test('debe de sacar el usuario por que el id no existe ', () => {
        const wrapper = shallowMount( EntryView, {
            props: {
                id: 'Este ID No existe en el store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
        expect( mockRouter.push ).toHaveBeenCalledWith({name: 'no-entry'})
    });

    test('debe de mostrar la entrada correctamente', () => {
        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
    });

    test('debe de borrar la entrada y salir',(done) => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }))
        wrapper.find('.btn-danger').trigger('click')
        expect( Swal.fire).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro!'
        })
        setTimeout( () =>{
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-NH7uKXKp-qWoAlduDNp')
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        }, 2)
        
    });
});