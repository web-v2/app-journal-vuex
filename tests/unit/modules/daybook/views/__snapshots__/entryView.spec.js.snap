// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div>
  <div class="entry-title d-flex justify-content-between p-2">
    <div><span class="text-success fs-3 fw-bold">17</span><span class="mx-1 fs-3">Noviembre</span><span class="mx-2 fs-4 fw-light">2022, Jueves</span></div>
    <div><input type="file" name="carga" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2">Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary">Subir foto <i class="fa fa-upload"></i></button></div>
  </div>
  <hr>
  <div class="d-flex flex-column px-3 h-75"><textarea name="post" id="post" cols="10" rows="3" placeholder="¿Qué sucedió hoy?"></textarea></div>
</div>
<img src="https://res.cloudinary.com/db2h2aoif/image/upload/v1668745453/vuejs/app-journal/Samir_V2_uu1haw.png" alt="entry-pinture" class="img thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
